### PR TITLE
Core/Misc: Sending player's own auras before all visible objects.

### LINKED ISSUE #15159: [3.3.5] Can't see other player's armor
Hello guys, I found this bug today, I saw sometimes that TrinityCore is not displaying player armor. I tried testing this on other cores but only TrinityCore have this bug. I don't know, it is newest revision, newest TDB, fresh compiled on Ubuntu. 

items

### PR DESCRIPTION
**Pull request by r4d1sh**
This should finally solve the problem of naked characters.

Many thanks to [xvwyh](https://github.com/xvwyh) for [explaining](https://github.com/TrinityCore/TrinityCore/issues/15159#issuecomment-1234988734) cause of this bug

**Issues addressed:**

Closes #15159

### CODE REVIEW

File: src/server/game/Maps/Map.cpp [hunk 1]
```diff
@@ -2950,13 +2951,23 @@ void Map::SendInitSelf(Player* player)
...
+    // build and send self update packet before sending to player his own auras
+    data.BuildPacket(&packet);
+    player->SendDirectMessage(&packet);
+
+    // send to player his own auras (this is needed here for timely initialization of some fields on client)
+    player->SendAurasForTarget(player);
+
+    // clean buffers for further work
+    packet.clear();
+    data.Clear();
```
Comment (leviathonbeast) [inline src/server/game/Maps/Map.cpp:2963]: is this a codestyle typo?

Comment (r4d1sh) [reply src/server/game/Maps/Map.cpp:2963]: > is this a codestyle typo?

what you mean?

Comment (leviathonbeast) [reply src/server/game/Maps/Map.cpp:2963]: +    packet.clear(); < not capital letter
+    data.Clear(); < capital letter

Comment (r4d1sh) [reply src/server/game/Maps/Map.cpp:2963]: > * packet.clear(); < not capital letter
> * data.Clear(); < capital letter

No, it's just that methods in different classes are called differently.
`ByteBuffer::clear()` and `UpdateData::Clear()`

Comment (leviathonbeast) [reply src/server/game/Maps/Map.cpp:2963]: never mind then. it just looks untidy to me :P

**Comment by jackpoz** [comment]
was this tested ingame and verified that the fields in the client are initialized ?

**Comment by CraftedRO** [comment]
31e7516b83aa Does it build, still hard to test and reproduce the issue even before fix.

**Comment by Aokromes** [comment]
> [31e7516](https://github.com/TrinityCore/TrinityCore/commit/31e7516b83aa84685b1f1cbbce2eb3bb24695585) Does it build, still hard to test and reproduce the issue even before fix.

it's not hard at all to reproduce. login with multiple characters on same location, don't move them and you will see few naked :)

**Comment by CraftedRO** [comment]
Ok tested multiple times and I let the character in initial position and it seems this doesn't fix the problem:

Before:

https://user-images.githubusercontent.com/24683355/204131896-7994cca7-ae11-499c-98bd-0e004d5d9bfd.mp4

After:

https://user-images.githubusercontent.com/24683355/204131908-16d97835-977c-4b39-8b9b-733ea855ad8c.mp4

It might be because this line is deleted here ?
`2942    WorldPacket packet;`

**Comment by CraftedRO** [comment]
Ok tested multiple times and I let the character in initial position and it seems this doesn't fix the problem:

Before:

https://user-images.githubusercontent.com/24683355/204131896-7994cca7-ae11-499c-98bd-0e004d5d9bfd.mp4

After:

https://user-images.githubusercontent.com/24683355/204131908-16d97835-977c-4b39-8b9b-733ea855ad8c.mp4

**Comment by jackpoz** [comment]
I would say one will have to debug WoW client to make sure the required fields are initialized

**Comment by jackpoz** [comment]
should this PR be marked as Draft as it doesn't solve the issue yet ?

**Comment by Kaev** [comment]
According to robinsch of our modding discord the naked character bug exists because the update object data of the character is just sent too early. Is there a clean way to delay that?

**Comment by xvwyh** [comment]
Oi, I wasn't notified of this because I was mentioned via a URL instead of @xvwyh.

@jackpoz, can you test this with characters that have some visible buffs on them? Because:
https://github.com/TrinityCore/TrinityCore/blob/b6d59f3f8885a382219a1e78fe85cfca281f1231/src/server/game/Entities/Player/Player.cpp#L23210-L23213
If the player that's logging in (the one that will see everyone else naked) doesn't have any visible auras on them - no packets will be sent and the memory will remain uninitialized. And you are testing with a brand new character that's unlikely to have any buffs. You'd need to add something like
```diff
-void Player::SendAurasForTarget(Unit* target) const
+void Player::SendAurasForTarget(Unit* target, bool force) const
 {
-     if (!target || target->GetVisibleAuras().empty())                  // speedup things 
+     if (!target || !force && target->GetVisibleAuras().empty())                  // speedup things 
```
And call it with force=true on login.
```diff
     // send to player his own auras (this is needed here for timely initialization of some fields on client)
-    player->SendAurasForTarget(player);
+    player->SendAurasForTarget(player, true);
```

Other than that, this PR should be functionally identical to the fix I made to my server. It vouch that it should work, barring any further differences between the codebases. Apologies for not mentioning this in my original message, I'll edit it now.

**Comment by r4d1sh** [comment]
I'd be grateful if someone could test this. Because I have no idea how to make sure the bug is no longer reproduced.

**Comment by Aokromes** [comment]
> I'd be grateful if someone could test this. Because I have no idea how to make sure the bug is no longer reproduced.

https://github.com/TrinityCore/TrinityCore/pull/28450#issuecomment-1328217322

**Comment by r4d1sh** [comment]
> > I'd be grateful if someone could test this. Because I have no idea how to make sure the bug is no longer reproduced.
> 
> [#28450 (comment)](https://github.com/TrinityCore/TrinityCore/pull/28450#issuecomment-1328217322)

Ok I use five characters in one place for testing. Two of them were created earlier and three right now. One of them had a visible aura, the rest did not. I made several attempts and every time in all open clients I saw all the characters dressed.

So it seems to me that everything works.

<details>
<summary>Screenshot</summary>

image
</details>

**Comment by Faq** [comment]
This actually reminds me of similar bug what other players sees different, but with mounted npcs and their relocation https://github.com/TrinityCore/TrinityCore/issues/20136#issuecomment-687663047

**Comment by CraftedRO** [comment]
yea, now problem is solved as I coud test multiple times with same characters and same positions.

**Comment by jackpoz** [comment]
is this PR ready to be merged then ?

**Comment by r4d1sh** [comment]
> is this PR ready to be merged then ?

Yes

**Comment by jackpoz** [comment]
Thanks for the PR 🚀 😃

**Pull request merged**